{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 169,
   "id": "68a8de9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import bmt\n",
    "import requests\n",
    "import linkml\n",
    "from linkml_runtime import SchemaView\n",
    "from collections import defaultdict\n",
    "from typing import Tuple, List\n",
    "import requests\n",
    "import yaml\n",
    "from pprint import pprint\n",
    "from biothings_explorer.smartapi_kg.dataload import load_specs\n",
    "import pandas\n",
    "import json\n",
    "\n",
    "tk = bmt.Toolkit('https://raw.githubusercontent.com/biolink/biolink-model/2.2.12/biolink-model.yaml')\n",
    "tsv_file = open(\"predicates.tsv\", \"w\")\n",
    "tsv_attributes = open(\"attributes.tsv\", \"w\")\n",
    "tsv_writer = csv.writer(tsv_file, delimiter='\\t')\n",
    "tsv_writer_att = csv.writer(tsv_attributes, delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "e56dc0d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def sample_predicates():\n",
    "    specs = load_specs()\n",
    "    kp_titles = []\n",
    "    for spec in specs:\n",
    "        if 'x-translator' not in spec['info']:\n",
    "            continue\n",
    "        if spec['info']['x-translator']['component'] == 'KP':\n",
    "            kp_titles.append(spec['info']['title'])\n",
    "        if 'servers' not in spec:\n",
    "            continue\n",
    "        else:\n",
    "            url = spec['servers'][0]['url']\n",
    "            apititle = '_'.join(spec['info']['title'].split())\n",
    "            if url.endswith('/'):\n",
    "                url = url[:-1]\n",
    "            predicates_url = f'{url}/meta_knowledge_graph'\n",
    "            print(predicates_url)\n",
    "            trapi, predicates = get_predicates(predicates_url)\n",
    "            # if is_trapi:\n",
    "            if not predicates:\n",
    "                continue\n",
    "            else:\n",
    "                print(\"got predicates\")\n",
    "                preds, attribs, url = dump_trapi_predicate_results(predicates, predicates_url)\n",
    "                predicates = set(preds)\n",
    "                attributes = set(attribs)\n",
    "                for pred in predicates:\n",
    "                    tsv_writer.writerow([apititle, url, pred])\n",
    "                for attrib in attributes:\n",
    "                    tsv_writer_att.writerow([apititle, url, attrib])\n",
    "            # else:\n",
    "            #     dump_smartapi_predicate_results(spec['info']['title'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "01040319",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def in_biolink_model(predicate):\n",
    "    is_predicate = tk.is_predicate(predicate)\n",
    "    return is_predicate\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "09086adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def dump_trapi_predicate_results(predicates, url):\n",
    "    preds = []\n",
    "    attribs = []\n",
    "    for edge in predicates.get('edges'):\n",
    "        predicate = edge.get('predicate')\n",
    "        subject = edge.get('subject')\n",
    "        tobject = edge.get('object')\n",
    "        if 'attributes' in edge and edge.get('attributes') is not None:\n",
    "            for attribute in edge.get('attributes'):\n",
    "                attribs.append(attribute.get('attribute_type_id'))\n",
    "        preds.append(predicate)\n",
    "    return preds, attribs, url\n",
    "\n",
    "            # tsv_writer.writerow([subject, predicate, tobject, url])\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "86b86b79",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_predicates(pr_url):\n",
    "    try:\n",
    "        response = requests.get(pr_url)\n",
    "        if response.status_code == 200:\n",
    "            return True, response.json()\n",
    "        else:\n",
    "            return False, {}\n",
    "    except:\n",
    "        return False, {}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "383c74f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/socio_environmental_exposures_api/v1/meta_knowledge_graph\n",
      "/cmaq_exposures_api/v1/meta_knowledge_graph\n",
      "/roadway_proximity_api/v1/meta_knowledge_graph\n",
      "https://name-resolution-sri.renci.org/meta_knowledge_graph\n",
      "https://icees-pcd.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://icees-dili-dev.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://www.ebi.ac.uk/proteins/api/meta_knowledge_graph\n",
      "https://icees-asthma-dev.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://icees-asthma.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://translator-workflow-runner.renci.org/meta_knowledge_graph\n",
      "https://icees-pcd-dev.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://www.ebi.ac.uk/ols/api/meta_knowledge_graph\n",
      "https://www.ncbi.nlm.nih.gov/research/bionlp/litvar/api/v1/meta_knowledge_graph\n",
      "https://icees-dili.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://www.ebi.ac.uk/QuickGO/services/meta_knowledge_graph\n",
      "https://ontology-kp.apps.renci.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://trapi.monarchinitiative.org/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://medikanren-trapi.ci.transltr.io/v2/meta_knowledge_graph\n",
      "https://automat.renci.org/gtex/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/intact/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/gtopdb/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/drugcentral/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/chemical-normalization/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/ctd/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/cord19/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/panther/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/textminingkp/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://bl-lookup-sri.renci.org/meta_knowledge_graph\n",
      "https://automat.renci.org/hmdb/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/covidkopkg/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://nodenormalization-sri.renci.org/1.2/meta_knowledge_graph\n",
      "https://nodenormalization-dev.apps.renci.org/1.2/meta_knowledge_graph\n",
      "https://automat.renci.org/pharos/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/human-goa/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/uberongraph/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/biolink/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/viral-proteome/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://cam-kp-api.renci.org/1.2.0/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/hgnc/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://cam-kp-api.renci.org/1.2.0/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/robokopkg/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/ontological-hierarchy/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/foodb/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/gwas-catalog/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/molepro-fda/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://automat.renci.org/hetio/1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://explanatory-agent.azurewebsites.net/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/uberon/meta_knowledge_graph\n",
      "https://aragorn-dev.apps.renci.org/1.2/meta_knowledge_graph\n",
      "http://chp.thayer.dartmouth.edu/meta_knowledge_graph\n",
      "got predicates\n",
      "https://smart-api.info/api/meta_knowledge_graph\n",
      "https://biothings.ncats.io/hpo/meta_knowledge_graph\n",
      "https://aragorn.renci.org/1.2/meta_knowledge_graph\n",
      "https://cooccurrence-p36smkc6hq-uc.a.run.app/meta_knowledge_graph\n",
      "got predicates\n",
      "https://spokekp.healthdatascience.cloud/api/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://cohd.io/api/meta_knowledge_graph\n",
      "got predicates\n",
      "https://openpredict.semanticscience.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://mygene.info/v3/meta_knowledge_graph\n",
      "https://myvariant.info/v1/meta_knowledge_graph\n",
      "https://biothings.ncats.io/rhea/meta_knowledge_graph\n",
      "https://molepro-trapi.transltr.io/molepro/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://ia.healthdatascience.cloud/api/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "http://mydisease.info/v1/meta_knowledge_graph\n",
      "https://translator.broadinstitute.org/genetics_provider/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/DISEASES/meta_knowledge_graph\n",
      "https://translator.broadinstitute.org/genetics_provider/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://translator.broadinstitute.org/genetics_provider/trapi/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/mgigene2phenotype/meta_knowledge_graph\n",
      "https://mychem.info/v1/meta_knowledge_graph\n",
      "https://biothings.ncats.io/ebigene2phenotype/meta_knowledge_graph\n",
      "https://biothings.ncats.io/go_bp/meta_knowledge_graph\n",
      "https://arax.ncats.io/beta/api/rtxkg2/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/go_cc/meta_knowledge_graph\n",
      "https://arax.ncats.io/api/rtxkg2/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://api.bte.ncats.io/v1/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/idisk/meta_knowledge_graph\n",
      "https://biothings.ncats.io/go_mf/meta_knowledge_graph\n",
      "https://api.bte.ncats.io/v1/team/Service%20Provider/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/multiomics_wellness_kp/meta_knowledge_graph\n",
      "https://api.collaboratory.semanticscience.org/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/dgidb/meta_knowledge_graph\n",
      "https://biothings.ncats.io/semmeddb/meta_knowledge_graph\n",
      "https://biothings.ncats.io/drug_response_kp/meta_knowledge_graph\n",
      "https://api.monarchinitiative.org/api/meta_knowledge_graph\n",
      "https://biothings.ncats.io/text_mining_targeted_association/meta_knowledge_graph\n",
      "https://arax.ncats.io/api/arax/v1.2/meta_knowledge_graph\n",
      "got predicates\n",
      "https://biothings.ncats.io/clinical_risk_kp/meta_knowledge_graph\n",
      "https://arax.ncats.io/beta/api/arax/v1.2/meta_knowledge_graph\n",
      "got predicates\n"
     ]
    }
   ],
   "source": [
    "\n",
    "if __name__ == '__main__':\n",
    "    sample_predicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "d26f6993",
   "metadata": {},
   "outputs": [],
   "source": [
    "sv = SchemaView('https://raw.githubusercontent.com/biolink/biolink-model/master/biolink-model.yaml')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "15087bc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('output.tsv', 'w', newline='') as f_output:\n",
    "    tsv_output = csv.writer(f_output, delimiter='\\t')\n",
    "\n",
    "    for s in sv.all_slots().values():\n",
    "        rows = []\n",
    "        if 'related to' in sv.slot_ancestors(s.name):\n",
    "            rows.append(s.name)\n",
    "            for x in sv.slot_children(s.name):\n",
    "                rows.append(x)\n",
    "\n",
    "            tsv_output.writerow(rows)\n",
    "            # else:\n",
    "            rows = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f6b57ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "for s in sv.all_slots().values():\n",
    "    if 'related to' in sv.slot_ancestors(s.name):\n",
    "        rows.append(s.name)\n",
    "        for child in sv.slot_children(s.name):\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "id": "78679f24",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_predicate_tree_data(biolink_version: str) -> Tuple[List[dict], str]:\n",
    "    # Grab Biolink yaml file and load into dictionary tree structures\n",
    "    response = requests.get(f\"https://raw.githubusercontent.com/biolink/biolink-model/\"\n",
    "                            f\"{biolink_version if biolink_version else 'master'}/biolink-model.yaml\",\n",
    "                            timeout=10)\n",
    "    if response.status_code == 200:\n",
    "        # Build predicates tree\n",
    "        biolink_model = yaml.safe_load(response.text)\n",
    "        parent_to_child_dict = defaultdict(set)\n",
    "        for slot_name_english, info in biolink_model[\"slots\"].items():\n",
    "            slot_name = convert_predicate_to_trapi_format(slot_name_english)\n",
    "            parent_name_english = info.get(\"is_a\")\n",
    "            if parent_name_english:\n",
    "                parent_name = convert_predicate_to_trapi_format(parent_name_english)\n",
    "                parent_to_child_dict[parent_name].add(slot_name)\n",
    "        root_node = {\"name\": \"related_to\"}\n",
    "        predicate_tree = get_tree_node_recursive(root_node, parent_to_child_dict)\n",
    "        biolink_version = biolink_model[\"version\"]\n",
    "        return [predicate_tree], biolink_version\n",
    "    else:\n",
    "        return [], \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "id": "fb247a87",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_tree_node_recursive(root_node: dict, parent_to_child_map: dict):\n",
    "    root_name = root_node[\"name\"]\n",
    "    children_predicates = parent_to_child_map.get(root_name, [])\n",
    "    if children_predicates:\n",
    "        children = []\n",
    "        for child_predicate in children_predicates:\n",
    "            child_node = {\"name\": child_predicate, \"parent\": root_name}\n",
    "            child_node = get_tree_node_recursive(child_node, parent_to_child_map)\n",
    "            children.append(child_node)\n",
    "        root_node[\"children\"] = sorted(children, key=lambda x: x[\"name\"])\n",
    "    return root_node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "2961b6da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['related_to',\n",
      " 'composed_primarily_of',\n",
      " 'disease_has_location',\n",
      " 'related_to_at_concept_level',\n",
      " 'broad_match',\n",
      " 'close_match',\n",
      " 'exact_match',\n",
      " 'same_as',\n",
      " 'narrow_match',\n",
      " 'subclass_of',\n",
      " 'superclass_of',\n",
      " 'related_to_at_instance_level',\n",
      " 'active_in',\n",
      " 'acts_upstream_of',\n",
      " 'acts_upstream_of_negative_effect',\n",
      " 'acts_upstream_of_or_within',\n",
      " 'acts_upstream_of_or_within_negative_effect',\n",
      " 'acts_upstream_of_or_within_positive_effect',\n",
      " 'acts_upstream_of_positive_effect',\n",
      " 'affected_by',\n",
      " 'abundance_affected_by',\n",
      " 'abundance_decreased_by',\n",
      " 'abundance_increased_by',\n",
      " 'activity_affected_by',\n",
      " 'activity_decreased_by',\n",
      " 'activity_increased_by',\n",
      " 'degradation_affected_by',\n",
      " 'degradation_decreased_by',\n",
      " 'degradation_increased_by',\n",
      " 'disrupted_by',\n",
      " 'entity_regulated_by_entity',\n",
      " 'entity_negatively_regulated_by_entity',\n",
      " 'entity_positively_regulated_by_entity',\n",
      " 'expression_affected_by',\n",
      " 'expression_decreased_by',\n",
      " 'expression_increased_by',\n",
      " 'folding_affected_by',\n",
      " 'folding_decreased_by',\n",
      " 'folding_increased_by',\n",
      " 'is_ameliorated_by',\n",
      " 'is_exacerbated_by',\n",
      " 'localization_affected_by',\n",
      " 'localization_decreased_by',\n",
      " 'localization_increased_by',\n",
      " 'metabolic_processing_affected_by',\n",
      " 'metabolic_processing_decreased_by',\n",
      " 'metabolic_processing_increased_by',\n",
      " 'molecular_modification_affected_by',\n",
      " 'molecular_modification_decreased_by',\n",
      " 'molecular_modification_increased_by',\n",
      " 'mutation_rate_affected_by',\n",
      " 'mutation_rate_decreased_by',\n",
      " 'mutation_rate_increased_by',\n",
      " 'process_regulated_by_process',\n",
      " 'process_negatively_regulated_by_process',\n",
      " 'process_positively_regulated_by_process',\n",
      " 'response_affected_by',\n",
      " 'response_decreased_by',\n",
      " 'response_increased_by',\n",
      " 'secretion_affected_by',\n",
      " 'secretion_decreased_by',\n",
      " 'secretion_increased_by',\n",
      " 'splicing_affected_by',\n",
      " 'splicing_decreased_by',\n",
      " 'splicing_increased_by',\n",
      " 'stability_affected_by',\n",
      " 'stability_decreased_by',\n",
      " 'stability_increased_by',\n",
      " 'synthesis_affected_by',\n",
      " 'synthesis_decreased_by',\n",
      " 'synthesis_increased_by',\n",
      " 'transport_affected_by',\n",
      " 'transport_decreased_by',\n",
      " 'transport_increased_by',\n",
      " 'uptake_affected_by',\n",
      " 'uptake_decreased_by',\n",
      " 'uptake_increased_by',\n",
      " 'affects',\n",
      " 'affects_abundance_of',\n",
      " 'decreases_abundance_of',\n",
      " 'increases_abundance_of',\n",
      " 'affects_activity_of',\n",
      " 'decreases_activity_of',\n",
      " 'increases_activity_of',\n",
      " 'affects_degradation_of',\n",
      " 'decreases_degradation_of',\n",
      " 'increases_degradation_of',\n",
      " 'affects_expression_in',\n",
      " 'affects_expression_of',\n",
      " 'decreases_expression_of',\n",
      " 'increases_expression_of',\n",
      " 'affects_folding_of',\n",
      " 'decreases_folding_of',\n",
      " 'increases_folding_of',\n",
      " 'affects_localization_of',\n",
      " 'decreases_localization_of',\n",
      " 'increases_localization_of',\n",
      " 'affects_metabolic_processing_of',\n",
      " 'decreases_metabolic_processing_of',\n",
      " 'increases_metabolic_processing_of',\n",
      " 'affects_molecular_modification_of',\n",
      " 'decreases_molecular_modification_of',\n",
      " 'increases_molecular_modification_of',\n",
      " 'affects_mutation_rate_of',\n",
      " 'decreases_mutation_rate_of',\n",
      " 'increases_mutation_rate_of',\n",
      " 'affects_response_to',\n",
      " 'decreases_response_to',\n",
      " 'increases_response_to',\n",
      " 'affects_secretion_of',\n",
      " 'decreases_secretion_of',\n",
      " 'increases_secretion_of',\n",
      " 'affects_splicing_of',\n",
      " 'decreases_splicing_of',\n",
      " 'increases_splicing_of',\n",
      " 'affects_stability_of',\n",
      " 'decreases_stability_of',\n",
      " 'increases_stability_of',\n",
      " 'affects_synthesis_of',\n",
      " 'decreases_synthesis_of',\n",
      " 'increases_synthesis_of',\n",
      " 'affects_transport_of',\n",
      " 'decreases_transport_of',\n",
      " 'increases_transport_of',\n",
      " 'affects_uptake_of',\n",
      " 'decreases_uptake_of',\n",
      " 'increases_uptake_of',\n",
      " 'ameliorates',\n",
      " 'treats',\n",
      " 'disrupts',\n",
      " 'entity_regulates_entity',\n",
      " 'entity_negatively_regulates_entity',\n",
      " 'entity_positively_regulates_entity',\n",
      " 'exacerbates',\n",
      " 'process_regulates_process',\n",
      " 'process_negatively_regulates_process',\n",
      " 'process_positively_regulates_process',\n",
      " 'affects_risk_for',\n",
      " 'predisposes',\n",
      " 'prevents',\n",
      " 'associated_with',\n",
      " 'associated_with_resistance_to',\n",
      " 'associated_with_sensitivity_to',\n",
      " 'correlated_with',\n",
      " 'biomarker_for',\n",
      " 'coexpressed_with',\n",
      " 'has_biomarker',\n",
      " 'negatively_correlated_with',\n",
      " 'occurs_together_in_literature_with',\n",
      " 'positively_correlated_with',\n",
      " 'resistance_associated_with',\n",
      " 'sensitivity_associated_with',\n",
      " 'caused_by',\n",
      " 'coexists_with',\n",
      " 'colocalizes_with',\n",
      " 'in_cell_population_with',\n",
      " 'in_complex_with',\n",
      " 'in_pathway_with',\n",
      " 'completed_by',\n",
      " 'contains_process',\n",
      " 'contraindicated_for',\n",
      " 'contributes_to',\n",
      " 'causes',\n",
      " 'adverse_event_caused_by',\n",
      " 'causes_adverse_event',\n",
      " 'contribution_from',\n",
      " 'contributor',\n",
      " 'author',\n",
      " 'editor',\n",
      " 'provider',\n",
      " 'publisher',\n",
      " 'decreased_amount_in',\n",
      " 'decreases_amount_or_activity_of',\n",
      " 'derives_from',\n",
      " 'is_metabolite_of',\n",
      " 'derives_into',\n",
      " 'has_metabolite',\n",
      " 'develops_from',\n",
      " 'develops_into',\n",
      " 'diagnoses',\n",
      " 'disease_has_basis_in',\n",
      " 'gene_product_of',\n",
      " 'genetic_association',\n",
      " 'condition_associated_with_gene',\n",
      " 'gene_associated_with_condition',\n",
      " 'has_completed',\n",
      " 'has_contraindication',\n",
      " 'has_contributor',\n",
      " 'has_author',\n",
      " 'has_editor',\n",
      " 'has_provider',\n",
      " 'has_publisher',\n",
      " 'has_decreased_amount',\n",
      " 'has_gene_product',\n",
      " 'has_increased_amount',\n",
      " 'has_manifestation',\n",
      " 'has_molecular_consequence',\n",
      " 'has_not_completed',\n",
      " 'has_participant',\n",
      " 'actively_involves',\n",
      " 'has_capability',\n",
      " 'enabled_by',\n",
      " 'has_catalyst',\n",
      " 'has_input',\n",
      " 'consumes',\n",
      " 'has_output',\n",
      " 'has_substrate',\n",
      " 'has_phenotype',\n",
      " 'has_real_world_evidence_of_association_with',\n",
      " 'has_sequence_location',\n",
      " 'has_sequence_variant',\n",
      " 'has_frameshift_variant',\n",
      " 'has_missense_variant',\n",
      " 'has_nearby_variant',\n",
      " 'has_non_coding_variant',\n",
      " 'has_nonsense_variant',\n",
      " 'has_splice_site_variant',\n",
      " 'has_synonymous_variant',\n",
      " 'has_target',\n",
      " 'has_upstream_actor',\n",
      " 'has_negative_upstream_actor',\n",
      " 'has_positive_upstream_actor',\n",
      " 'has_upstream_or_within_actor',\n",
      " 'has_negative_upstream_or_within_actor',\n",
      " 'has_positive_upstream_or_within_actor',\n",
      " 'in_linkage_disequilibrium_with',\n",
      " 'in_taxon',\n",
      " 'increased_amount_of',\n",
      " 'increases_amount_or_activity_of',\n",
      " 'interacts_with',\n",
      " 'directly_interacts_with',\n",
      " 'genetically_interacts_with',\n",
      " 'physically_interacts_with',\n",
      " 'chemically_interacts_with',\n",
      " 'molecularly_interacts_with',\n",
      " 'is_diagnosed_by',\n",
      " 'is_molecular_consequence_of',\n",
      " 'is_sequence_variant_of',\n",
      " 'is_frameshift_variant_of',\n",
      " 'is_missense_variant_of',\n",
      " 'is_nearby_variant_of',\n",
      " 'is_non_coding_variant_of',\n",
      " 'is_nonsense_variant_of',\n",
      " 'is_splice_site_variant_of',\n",
      " 'is_synonymous_variant_of',\n",
      " 'lacks_part',\n",
      " 'located_in',\n",
      " 'expressed_in',\n",
      " 'location_of',\n",
      " 'expresses',\n",
      " 'manifestation_of',\n",
      " 'mentioned_by',\n",
      " 'mentions',\n",
      " 'missing_from',\n",
      " 'model_of',\n",
      " 'models',\n",
      " 'not_completed_by',\n",
      " 'occurs_in',\n",
      " 'occurs_in_disease',\n",
      " 'opposite_of',\n",
      " 'overlaps',\n",
      " 'has_part',\n",
      " 'has_active_ingredient',\n",
      " 'has_excipient',\n",
      " 'has_food_component',\n",
      " 'has_plasma_membrane_part',\n",
      " 'has_variant_part',\n",
      " 'part_of',\n",
      " 'food_component_of',\n",
      " 'is_active_ingredient_of',\n",
      " 'is_excipient_of',\n",
      " 'plasma_membrane_part_of',\n",
      " 'variant_part_of',\n",
      " 'participates_in',\n",
      " 'actively_involved_in',\n",
      " 'capable_of',\n",
      " 'catalyzes',\n",
      " 'enables',\n",
      " 'is_input_of',\n",
      " 'consumed_by',\n",
      " 'is_output_of',\n",
      " 'is_substrate_of',\n",
      " 'phenotype_of',\n",
      " 'produced_by',\n",
      " 'produces',\n",
      " 'related_condition',\n",
      " 'risk_affected_by',\n",
      " 'prevented_by',\n",
      " 'sequence_location_of',\n",
      " 'similar_to',\n",
      " 'chemically_similar_to',\n",
      " 'homologous_to',\n",
      " 'orthologous_to',\n",
      " 'paralogous_to',\n",
      " 'xenologous_to',\n",
      " 'target_for',\n",
      " 'taxon_of',\n",
      " 'temporally_related_to',\n",
      " 'preceded_by',\n",
      " 'precedes',\n",
      " 'transcribed_from',\n",
      " 'transcribed_to',\n",
      " 'translates_to',\n",
      " 'translation_of',\n",
      " 'treated_by',\n",
      " 'approved_for_treatment_by']\n"
     ]
    }
   ],
   "source": [
    "with open('test.txt', 'a') as f:\n",
    "    row = []\n",
    "    ptd = load_predicate_tree_data(\"master\")\n",
    "    for item in ptd[0]:\n",
    "        f.write(item[\"name\"] + \",,,,,\" + \"\\n\")\n",
    "        row.append(item[\"name\"])\n",
    "        if item.get(\"children\"):\n",
    "            for child in item['children']:\n",
    "                f.write(\",\" + child['name'] + \",,,,\" + \"\\n\")\n",
    "                row.append(child['name'])\n",
    "                if child.get(\"children\"):\n",
    "                    for gchild in child['children']:\n",
    "                        f.write(\",,\" + gchild['name'] + \",,,\" + \"\\n\")\n",
    "                        row.append(gchild['name'])\n",
    "                        if gchild.get(\"children\"):\n",
    "                            for ggchild in gchild['children']:\n",
    "                                f.write(\",,,\" + ggchild['name'] + \",,\" + \"\\n\")\n",
    "                                row.append(ggchild['name'])\n",
    "                                if ggchild.get(\"children\"):\n",
    "                                    for gggchild in ggchild['children']:\n",
    "                                        f.write(\",,,,\" + gggchild['name'] + \",\" + \"\\n\")\n",
    "                                        row.append(gggchild['name'])\n",
    "        pprint(row)\n",
    "        rows = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "4c9d7d87",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_predicate_to_trapi_format(english_predicate: str) -> str:\n",
    "    # Converts a string like \"treated by\" to \"treated_by\"\n",
    "    return english_predicate.replace(' ', '_')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f6254bd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
